core/src/main/java/org/apache/druid/query/QueryException.java

```diff
@@ -30,12 +29,126 @@
 import java.util.function.Function;
 
 /**
- * Base serializable error response
- *
+ * Base serializable error response.
+ * <p>
+ * The Object Model that QueryException follows is a little non-intuitive as the primary way that a QueryException is
+ * generated is through a child class.  However, those child classes are *not* equivalent to a QueryException, instead
+ * they act as a Factory of QueryException objects.  This can be seen in two different places.
+ * <p>
+ * 1. When sanitize() is called, the response is a QueryException without any indication of which original exception
+ * occurred.
+ * 2. When these objects get serialized across the wire the recipient deserializes a QueryException. The client is
+ * never expected, and fundamentally is not allowed to, ever deserialize a child class of QueryException.
+ * <p>
+ * For this reason, QueryException must contain all potential state that any of its child classes could ever want to
+ * push across the wire.  Additionally, any catch clauses expecting one of the child Exceptions must know that it is
+ * running inside of code where the exception has not traveled across the wire.  If there is a chance that the
+ * exception could have been serialized across the wire, the code must catch a QueryException and check the errorCode
+ * instead.
+ * <p>
+ * As a correlary, adding new state or adjusting the logic of this class must always be done in a backwards-compatible
+ * fashion across all child classes of QueryException.
+ * <p>
+ * If there is any need to do different logic based on the type of error that has happened, the only reliable method
+ * of discerning the type of the error is to look at the errorCode String.  Because these Strings are considered part
+ * of the API, they are not allowed to change and must maintain their same semantics.  The known errorCode Strings
+ * are pulled together as public static fields on this class in order to make it more clear what options exist.
+ * <p>
  * QueryResource and SqlResource are expected to emit the JSON form of this object when errors happen.
  */
 public class QueryException extends RuntimeException implements SanitizableException
 {
+  /**
+   * Error codes
+   */
+  public static final String JSON_PARSE_ERROR_CODE = "Json parse failed";
+  public static final String BAD_QUERY_CONTEXT_ERROR_CODE = "Query context parse failed";
+  public static final String QUERY_CAPACITY_EXCEEDED_ERROR_CODE = "Query capacity exceeded";
+  public static final String QUERY_INTERRUPTED_ERROR_CODE = "Query interrupted";
+  // Note: the proper spelling is with a single "l", but the version with
```

Comment:
I left this comment because Gian pointed out the correct "cancelled" (2 l's) spelling, but we noted we can't change the constant because it is part of the public API.

But, this PR changes the constant (by moving it from one place to another), so we might as well fix the spelling of the constant name while we're breaking things.